### PR TITLE
feat: gzip S3/Admin API responses whenever applicable

### PR DESCRIPTION
**Pull request by klauspost**
## Description

Enable serverside gzip compression for all, but GetObject requests.

## Motivation and Context

Reduce network transfer and return results faster to clients on slow connections.

Disabled for HEAD requests [(ancient nginx problem)](https://trac.nginx.org/nginx/ticket/358), very small responses, range requests (which only GetObject supports), compressed content types and of course if not supported by the client.

Note that trace will still show the *uncompressed* size and the `httpTraceAll` will also show uncompressed content.

The compressor will not be engaged until the response body is starting to be written, so until that point requests will not take memory.

## How to test this PR?

`mc` and `warp` will require small updates.

mc:
```diff
diff --git a/cmd/client-s3.go b/cmd/client-s3.go
index f027d4a7..44d62154 100644
--- a/cmd/client-s3.go
+++ b/cmd/client-s3.go
@@ -37,9 +37,10 @@ import (
 	"sync"
 	"time"
 
+	"github.com/klauspost/compress/gzhttp"
 	"github.com/minio/mc/pkg/httptracer"
 	"github.com/minio/mc/pkg/probe"
-	minio "github.com/minio/minio-go/v7"
+	"github.com/minio/minio-go/v7"
 	"github.com/minio/minio-go/v7/pkg/credentials"
 	"github.com/minio/minio-go/v7/pkg/encrypt"
 	"github.com/minio/minio-go/v7/pkg/lifecycle"
@@ -184,7 +185,7 @@ func newFactory() func(config *Config) (Client, *probe.Error) {
 					// 	return nil, probe.NewError(e)
 					// }
 				}
-				transport = tr
+				transport = gzhttp.Transport(tr)
 			}
 
 			if config.Debug {
 ```
warp:
```
diff --git a/cli/client.go b/cli/client.go
index fb60aea..a11f5ff 100644
--- a/cli/client.go
+++ b/cli/client.go
@@ -30,6 +30,7 @@ import (
 	"sync"
 	"time"
 
+	"github.com/klauspost/compress/gzhttp"
 	"github.com/minio/cli"
 	"github.com/minio/madmin-go"
 	"github.com/minio/mc/pkg/probe"
@@ -211,7 +212,7 @@ func clientTransport(ctx *cli.Context) http.RoundTripper {
 		// See https://github.com/golang/go/issues/14275
 		http2.ConfigureTransport(tr)
 	}
-	return tr
+	return gzhttp.Transport(tr)
 }
 
 // parseHosts will parse the host parameter given.
```

Speed compare, before -> after:
```
-------------------
Operation: LIST
Operations: 983 -> 1018
* Average: +3.32% (+269.4) obj/s
* First Byte: Average: +3.237342ms (+0%), Median: +5.0007ms (+0%), Best: +56.5466ms (+6%), Worst: +94.0181ms (+7%)
* Fastest: +3.10% (+261.3) obj/s
* 50% Median: +3.12% (+255.1) obj/s
* Slowest: +4.61% (+354.0) obj/s
```
Within margin of error, I wouldn't expect it to actually be faster.

Size compare, of a regular listing. Real objects, not benchmark generated. Ignore speed since it was clearly cached on the second run. Trace modified to show compressed size.

```
//  DISABLED.............
2021-07-06T15:27:58:000 [200 OK] github.com/klauspost/compress/gzhttp.NewWrapper.func1.1 127.0.0.1:9001/testbucket/?continuation-token=c3JjMy9zcmMvbmV0L2h0dHAvaHR0cHV0aWwvcmV2ZXJzZXByb3h5LmdvW21pbmlvX2NhY2hlOnYyLGlkOjM3MzI3NWJhLWExYjUtNGJmZi04ZjY1LWU1ODgwN2ZjMDU0YyxwOjAsczowXQ%3D%3D&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=  127.0.0.1         1.085961s    ↑ 77 B ↓ 1.5 MiB
2021-07-06T15:27:59:000 [200 OK] github.com/klauspost/compress/gzhttp.NewWrapper.func1.1 127.0.0.1:9001/testbucket/?continuation-token=c3JjNC9zcmMvY21kL3ZlbmRvci9nb2xhbmcub3JnL3gvc3lzL3dpbmRvd3Mvemtub3duZm9sZGVyaWRzX3dpbmRvd3MuZ29bbWluaW9fY2FjaGU6djIsaWQ6MzczMjc1YmEtYTFiNS00YmZmLThmNjUtZTU4ODA3ZmMwNTRjLHA6MCxzOjBd&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=  127.0.0.1         1.593553s    ↑ 77 B ↓ 1.5 MiB
2021-07-06T15:28:01:000 [200 OK] github.com/klauspost/compress/gzhttp.NewWrapper.func1.1 127.0.0.1:9001/testbucket/?continuation-token=c3JjNS9zcmMvY21kL2NvbXBpbGUvaW50ZXJuYWwvZ2MvdGVzdGRhdGEvY2xvc3VyZV90ZXN0LmdvW21pbmlvX2NhY2hlOnYyLGlkOjM3MzI3NWJhLWExYjUtNGJmZi04ZjY1LWU1ODgwN2ZjMDU0YyxwOjAsczowXQ%3D%3D&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=  127.0.0.1         1.589316s    ↑ 77 B ↓ 1.5 MiB
2021-07-06T15:28:02:000 [200 OK] github.com/klauspost/compress/gzhttp.NewWrapper.func1.1 127.0.0.1:9001/testbucket/?continuation-token=c3JjNS9zcmMvb3MvZXhlY191bml4X3Rlc3QuZ29bbWluaW9fY2FjaGU6djIsaWQ6MzczMjc1YmEtYTFiNS00YmZmLThmNjUtZTU4ODA3ZmMwNTRjLHA6MCxzOjBd&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=  127.0.0.1         1.58685s     ↑ 77 B ↓ 1.5 MiB
2021-07-06T15:28:04:000 [200 OK] github.com/klauspost/compress/gzhttp.NewWrapper.func1.1 127.0.0.1:9001/testbucket/?continuation-token=c3JjNi9zcmMvY3J5cHRvL2NpcGhlci94b3JfdGVzdC5nb1ttaW5pb19jYWNoZTp2MixpZDozNzMyNzViYS1hMWI1LTRiZmYtOGY2NS1lNTg4MDdmYzA1NGMscDowLHM6MF0%3D&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=  127.0.0.1         64.891ms     ↑ 77 B ↓ 1.3 MiB

////////////// ENABLED ...............

2021-07-06T17:35:12:000 [200 OK] github.com/klauspost/compress/gzhttp.NewWrapper.func1.1 127.0.0.1:9001/testbucket/?continuation-token=c3JjMy9zcmMvYXJjaGl2ZS90YXIvdGVzdGRhdGEvaXNzdWUxMDk2OC50YXJbbWluaW9fY2FjaGU6djIsaWQ6MmMwYmVmZWQtYzRlNy00YzNiLTkzNzUtZTE3NDlhODhlMjEzLHA6MCxzOjBd&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=  127.0.0.1         99.744ms     ↑ 93 B ↓ 158 KiB
2021-07-06T17:35:12:000 [200 OK] github.com/klauspost/compress/gzhttp.NewWrapper.func1.1 127.0.0.1:9001/testbucket/?continuation-token=c3JjMy9zcmMvbmV0L2h0dHAvaHR0cHV0aWwvcmV2ZXJzZXByb3h5LmdvW21pbmlvX2NhY2hlOnYyLGlkOjJjMGJlZmVkLWM0ZTctNGMzYi05Mzc1LWUxNzQ5YTg4ZTIxMyxwOjAsczowXQ%3D%3D&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=  127.0.0.1         90.696ms     ↑ 93 B ↓ 158 KiB
2021-07-06T17:35:13:000 [200 OK] github.com/klauspost/compress/gzhttp.NewWrapper.func1.1 127.0.0.1:9001/testbucket/?continuation-token=c3JjNC9zcmMvY21kL3ZlbmRvci9nb2xhbmcub3JnL3gvc3lzL3dpbmRvd3Mvemtub3duZm9sZGVyaWRzX3dpbmRvd3MuZ29bbWluaW9fY2FjaGU6djIsaWQ6MmMwYmVmZWQtYzRlNy00YzNiLTkzNzUtZTE3NDlhODhlMjEzLHA6MCxzOjBd&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=  127.0.0.1         87.077ms     ↑ 93 B ↓ 156 KiB
2021-07-06T17:35:13:000 [200 OK] github.com/klauspost/compress/gzhttp.NewWrapper.func1.1 127.0.0.1:9001/testbucket/?continuation-token=c3JjNS9zcmMvY21kL2NvbXBpbGUvaW50ZXJuYWwvZ2MvdGVzdGRhdGEvY2xvc3VyZV90ZXN0LmdvW21pbmlvX2NhY2hlOnYyLGlkOjJjMGJlZmVkLWM0ZTctNGMzYi05Mzc1LWUxNzQ5YTg4ZTIxMyxwOjAsczowXQ%3D%3D&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=  127.0.0.1         100.984ms    ↑ 93 B ↓ 158 KiB
2021-07-06T17:35:13:000 [200 OK] github.com/klauspost/compress/gzhttp.NewWrapper.func1.1 127.0.0.1:9001/testbucket/?continuation-token=c3JjNS9zcmMvb3MvZXhlY191bml4X3Rlc3QuZ29bbWluaW9fY2FjaGU6djIsaWQ6MmMwYmVmZWQtYzRlNy00YzNiLTkzNzUtZTE3NDlhODhlMjEzLHA6MCxzOjBd&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=  127.0.0.1         86.983ms     ↑ 93 B ↓ 160 KiB
2021-07-06T17:35:13:000 [200 OK] github.com/klauspost/compress/gzhttp.NewWrapper.func1.1 127.0.0.1:9001/testbucket/?continuation-token=c3JjNi9zcmMvY3J5cHRvL2NpcGhlci94b3JfdGVzdC5nb1ttaW5pb19jYWNoZTp2MixpZDoyYzBiZWZlZC1jNGU3LTRjM2ItOTM3NS1lMTc0OWE4OGUyMTMscDowLHM6MF0%3D&delimiter=&encoding-type=url&fetch-owner=true&list-type=2&prefix=  127.0.0.1         68.852ms     ↑ 93 B ↓ 139 KiB
```

For listing, close to 10:1 compression.

## Types of changes
- [x] Optimization (provides speedup with no functional changes)

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :x: [more...](#a8f976db-4d1b-45b1-86b3-1671c093ce42) |
<a name="a8f976db-4d1b-45b1-86b3-1671c093ce42"></a>
### 12631-2f72332/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s1:30452
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/16) Running aws-sdk-go tests ... done in 1 seconds
(2/16) Running aws-sdk-java tests ... done in 1 seconds
(3/16) Running aws-sdk-php tests ... done in 42 seconds
(4/16) Running aws-sdk-ruby tests ... done in 3 seconds
(5/16) Running awscli tests ... done in 54 seconds
(6/16) Running healthcheck tests ... done in 0 seconds
(7/16) Running mc tests ... done in 18 seconds
(8/16) Running minio-dotnet tests ... done in 28 seconds
(9/16) Running minio-go tests ... done in 35 seconds
(10/16) Running minio-java tests ... done in 39 seconds
(11/16) Running minio-js tests ... done in 36 seconds
(12/16) Running minio-py tests ... FAILED in 58 seconds
{
  "name": "minio-py:test_list_objects_with_1001_files",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-62e7e1db-deca-4d1e-af15-424e45930f56",
    "object_name": "7d594958-876c-4c12-8a01-bdd1c019548e_0 ~ 7d594958-876c-4c12-8a01-bdd1c019548e_1000"
  },
  "message": "Listed no of objects (1000), does not match the expected no of objects (2000)",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1247, in test_list_objects_with_1001_files\n    _test_list_objects_api(bucket_name, no_of_created_files)\n  File \"/mint/run/core/minio-py/tests.py\", line 1135, in _test_list_objects_api\n    \"expected no of objects ({})\").format(no_of_files, expected_no))\nValueError: Listed no of objects (1000), does not match the expected no of objects (2000)\n",
  "duration": 38848
}
(12/16) Running s3cmd tests ... done in 1 minutes and 6 seconds
(13/16) Running s3select tests ... done in 3 seconds
(14/16) Running security tests ... done in 0 seconds
(15/16) Running versioning tests ... done in 0 seconds

Executed 15 out of 16 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :x: [more...](#b882ad4f-01e2-40e2-b291-7843b4835917) |
<a name="b882ad4f-01e2-40e2-b291-7843b4835917"></a>
### 12631-a1362c8/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s7:30302
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/16) Running aws-sdk-go tests ... done in 1 seconds
(2/16) Running aws-sdk-java tests ... done in 1 seconds
(3/16) Running aws-sdk-php tests ... done in 42 seconds
(4/16) Running aws-sdk-ruby tests ... done in 3 seconds
(5/16) Running awscli tests ... done in 53 seconds
(6/16) Running healthcheck tests ... done in 0 seconds
(7/16) Running mc tests ... done in 20 seconds
(8/16) Running minio-dotnet tests ... done in 28 seconds
(9/16) Running minio-go tests ... done in 45 seconds
(10/16) Running minio-java tests ... done in 41 seconds
(11/16) Running minio-js tests ... done in 35 seconds
(12/16) Running minio-py tests ... FAILED in 57 seconds
{
  "name": "minio-py:test_list_objects_with_1001_files",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-caac4d08-03bc-4303-9b4e-e46658f48bd9",
    "object_name": "921775f2-1575-4052-914e-3170d9482380_0 ~ 921775f2-1575-4052-914e-3170d9482380_1000"
  },
  "message": "Listed no of objects (1000), does not match the expected no of objects (2000)",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1247, in test_list_objects_with_1001_files\n    _test_list_objects_api(bucket_name, no_of_created_files)\n  File \"/mint/run/core/minio-py/tests.py\", line 1135, in _test_list_objects_api\n    \"expected no of objects ({})\").format(no_of_files, expected_no))\nValueError: Listed no of objects (1000), does not match the expected no of objects (2000)\n",
  "duration": 38955
}
(12/16) Running s3cmd tests ... done in 1 minutes and 4 seconds
(13/16) Running s3select tests ... done in 3 seconds
(14/16) Running security tests ... done in 0 seconds
(15/16) Running versioning tests ... done in 0 seconds

Executed 15 out of 16 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
@klauspost wouldn't this be S3 API incompatible?

**Comment by harshavardhana** [comment]
> @klauspost wouldn't this be S3 API incompatible?

I am curious how other SDKs will work here since it's a gzipped transport, and they may not have auto-extract on the transport layer - or am I reading this correctly?

**Review by harshavardhana — CHANGES_REQUESTED** [review]
More than S3/admin - I was thinking we should do this for internode communication @klauspost as that can help in reducing the bandwidth.

**Comment by klauspost** [comment]
@harshavardhana For WAN it would be a benefit (hence why only doing calls that are potentially remote), with LAN probably not so much. And we definitely should use something like S2 which is much lighter.

`WalkDir` ready sends compressed data. I guess the only data intensive work we have left is data transfer to and from disks. With encryption and/or compression enabled it will not compress, so I guess that leaves default setups.

**Comment by klauspost** [comment]
> I am curious how other SDKs will work here since it's a gzipped transport, and they may not have auto-extract on the transport layer - or am I reading this correctly?

If the client sets `Accept-Encoding: gzip`, it signals that it will transparently decompress data. That can be expected to be true for any http client.

**Comment by harshavardhana** [comment]
> If the client sets `Accept-Encoding: gzip`, it signals that it will transparently decompress data. That can be expected to be true for any http client.

So it is basically based on this? @klauspost such that if the client accepts `Accept-Encoding: gzip` or not?  if the client says it's `Accept-Encoding: deflate` this won't work?

**Comment by klauspost** [comment]
@harshavardhana Correct. `deflate` was implemented differently (wrong) in some browsers at some point in ancient history, so it hasn't really been widely used, since it wasn't safe. `gzip` should work transparently.

**Comment by harshavardhana** [comment]
> @harshavardhana Correct. `deflate` was implemented differently (wrong) in some browsers at some point in ancient history, so it hasn't really been widely used, since it wasn't safe. `gzip` should work transparently.

oh nice this is a very interesting feature to test out, i shall do this soon.

File: cmd/admin-router.go [hunk 1]
```diff
@@ -46,174 +48,184 @@ func registerAdminRouter(router *mux.Router, enableConfigOps, enableIAMOps bool)
...
 
 		if !globalIsGateway {
 			// Keep obdinfo for backward compatibility with mc
 			adminRouter.Methods(http.MethodGet).Path(adminVersion + "/obdinfo").
-				HandlerFunc(httpTraceHdrs(adminAPI.HealthInfoHandler))
+				HandlerFunc(gz(httpTraceHdrs(adminAPI.HealthInfoHandler)))
 			// -- Health API --
 			adminRouter.Methods(http.MethodGet).Path(adminVersion + "/healthinfo").
-				HandlerFunc(httpTraceHdrs(adminAPI.HealthInfoHandler))
+				HandlerFunc(gz(httpTraceHdrs(adminAPI.HealthInfoHandler)))
```
Comment (harshavardhana) [inline cmd/admin-router.go:225]: isn't this already gzipped? in some form on `mc`?  also this uses Trailers - does this work fine `mc admin subnet health` ?

File: cmd/api-router.go [hunk 2]
```diff
@@ -218,218 +231,218 @@ func registerAPIRouter(router *mux.Router) {
 		// Object operations
 		// HeadObject
 		router.Methods(http.MethodHead).Path("/{object:.+}").HandlerFunc(
-			collectAPIStats("headobject", maxClients(httpTraceAll(api.HeadObjectHandler))))
+			collectAPIStats("headobject", maxClients(gz(httpTraceAll(api.HeadObjectHandler)))))
```
Comment (harshavardhana) [inline cmd/api-router.go:234]: shouldn't this be not set? for HEAD operations?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/admin-router.go (hunk 1 above)
Comment (klauspost) [reply cmd/admin-router.go:225]: It should detect content type (even if not set explicitly) and skip compression. Even if it doesn't it handles pre-compressed data at >6GB/s, so not too worried.

It should work transparently, even if content is already gzipped.

File: cmd/api-router.go (hunk 2 above)
Comment (klauspost) [reply cmd/api-router.go:234]: Yeah, it is pretty much a no-op for HEAD operations.

**Review by vadmeste — APPROVED** [review]
LGTM

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**